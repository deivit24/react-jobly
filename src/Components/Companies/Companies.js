import React, { useState, useEffect } from 'react';
import API from '../../API';
import Search from '../Search/Search';
import CompanyCard from '../Cards/CompanyCard';

const Companies = () => {
  const [companies, setCompanies] = useState([]);

  useEffect(() => {
    async function getCompanies() {
      let companies = await API.getCompanies();

      setCompanies(companies);
    }
    getCompanies();
  }, []);

  async function handleSearch(search) {
    let companies = await API.getCompanies(search);
    setCompanies(companies);
  }
  let none = <p>Sorry no results</p>;
  let found = companies.map((company) => (
    <div key={company.handle} className="col-md-3 mb-3">
      <CompanyCard item={company} />
    </div>
  ));
  return (
    <section className="container mt-5">
      <h3 className="mt-5 mb-3 text-center">Search for Companies </h3>
      <Search endpoint="companies" searchFor={handleSearch} />
      <div className="row mt-5">{companies.length ? found : none}</div>
    </section>
  );
};

export default Companies;
