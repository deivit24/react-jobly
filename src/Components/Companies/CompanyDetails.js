import React, { useState, useEffect, useContext } from 'react';
import { useParams } from 'react-router-dom';
import API from '../../API';
import CompanyCard from '../Cards/CompanyCard';
import JobCard from '../Cards/JobCard';
import UserContext from '../../Helpers/UserContext';
import Loading from '../../Helpers/Loading';

const CompanyDetails = () => {
  const { handle } = useParams();
  const { currentUser } = useContext(UserContext);

  const [company, setCompany] = useState(null);

  useEffect(() => {
    async function getCompanyAndJobs() {
      const { jobs } = currentUser;
      const c = await API.getCompany(handle);

      const jobsIDsAppliedTo = new Set(jobs.map((job) => job.id));

      c.jobs = c.jobs.map((job) => ({
        ...job,
        state: jobsIDsAppliedTo.has(job.id) ? 'applied' : null,
      }));

      setCompany(c);
    }

    getCompanyAndJobs();
  }, [handle, currentUser]);

  async function apply(idx) {
    if (company && Array.isArray(company.jobs) && idx < company.jobs.length) {
      let jobId = company.jobs[idx].id;
      let message = await API.applyToJob(jobId);
      setCompany((c) => {
        let newCompany = { ...c };
        newCompany.jobs = newCompany.jobs.map((job) =>
          job.id === jobId ? { ...job, state: message } : job
        );
        return newCompany;
      });
    }
  }

  if (!company) {
    return <Loading type="spin" color="#212aa5" />;
  }

  return (
    <div className="container">
      <div className="row mt-3">
        <div className="col-md-12 text-center my-3">
          <h1>{company.name}</h1>
          <p>Open Positions Available</p>
        </div>
        <div className="col-md-4">
          <CompanyCard item={company} />
        </div>
        <div className="col-md-8">
          {company.jobs.map((job, idx) => (
            <div className="mb-3" key={idx}>
              <JobCard idx={idx} handleApply={() => apply(idx)} item={job} />
            </div>
          ))}
        </div>
      </div>
    </div>
  );
};

export default CompanyDetails;
