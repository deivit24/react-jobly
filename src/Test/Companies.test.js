import React from 'react';
import { render } from '@testing-library/react';
import Companies from '../Components/Companies/Companies';
import CompanyDetails from '../Components/Companies/CompanyDetails';
import { MemoryRouter } from 'react-router';
import { UserProvider } from '../testUtils';

//Testing Companies and Company Details

describe('Testing Companies and Company Details', () => {
  it('it renders Companies without crashing', () => {
    render(
      <MemoryRouter>
        <Companies />
      </MemoryRouter>
    );
  });
  it('it renders CompanyDetails without crashing', () => {
    render(
      <MemoryRouter>
        <UserProvider>
          <CompanyDetails />
        </UserProvider>
      </MemoryRouter>
    );
  });
  it('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <Companies />
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('matches snapshot', () => {
    const { asFragment } = render(
      <MemoryRouter>
        <UserProvider>
          <CompanyDetails />
        </UserProvider>
      </MemoryRouter>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
