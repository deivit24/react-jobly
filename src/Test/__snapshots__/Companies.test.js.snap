// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Companies and Company Details matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="container mt-5"
  >
    <h3
      class="mt-5 mb-3 text-center"
    >
      Search for Companies 
    </h3>
    <div
      class="Search mb-4"
    >
      <form
        class="form-inline"
      >
        <input
          class="form-control form-control-lg flex-grow-1"
          name="search"
          placeholder="Enter search term.."
          value=""
        />
        <button
          class="btn btn-lg btn-primary"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
    <div
      class="row mt-5"
    >
      <p>
        Sorry no results
      </p>
    </div>
  </section>
</DocumentFragment>
`;

exports[`Testing Companies and Company Details matches snapshot 2`] = `
<DocumentFragment>
  <div>
    Loading...
  </div>
</DocumentFragment>
`;

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="container mt-5"
  >
    <h3
      class="mt-5 mb-3 text-center"
    >
      Search for Companies 
    </h3>
    <div
      class="Search mb-4"
    >
      <form
        class="form-inline"
      >
        <input
          class="form-control form-control-lg flex-grow-1"
          name="search"
          placeholder="Enter search term.."
          value=""
        />
        <button
          class="btn btn-lg btn-primary"
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
    <div
      class="row mt-5"
    >
      <p>
        Sorry no results
      </p>
    </div>
  </section>
</DocumentFragment>
`;
